kubectl run: deprecate unused / nonuseful flags

brianpursley opened the pull request:
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:

The --cascade flag, has no practical effect when used, so it is being deprecated.

The following flags, which are unused and ignored by kubectl run, 
have been deprecated:

--filename
--force
--grace-period
--kustomize
--recursive
--timeout
--wait

This PR hides them and sets a deprecation message, so that they can be removed in a future release (1 year = kubectl 1.29 if this makes it into kubectl 1.26).

#### Which issue(s) this PR fixes:
Fixes #108630

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
Deprecated the following kubectl run flags, which are ignored if set: --cascade, --filename, --force, --grace-period, --kustomize, --recursive, --timeout, --wait
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs

```

brianpursley commented:
/hold

brianpursley commented:
/retest

brianpursley commented:
Updated the PR to specify which version the flags will be removed in, after a one year deprecation period.

Also hid the deprecated flags, so they don't show in help anymore.  I *think* this is OK to do because they are no-ops, completely ignored if set.

If someone is using the flags, their command will not fail and they will get a deprecation message about the flag.

KnVerey commented:
/triage accepted
/priority important-longterm

KnVerey commented on staging/src/k8s.io/kubectl/pkg/cmd/run/run.go:
```diff
@@ -166,6 +166,20 @@ func NewCmdRun(f cmdutil.Factory, streams genericclioptions.IOStreams) *cobra.Co
+
+	// Deprecate and hide unused flags.
+	// These flags are being added to the run command by DeleteFlags to support pod deletion after attach,
+	// but they are not used if set, so they effectively do nothing.
```
It looks like the values are passed through during `Complete`, so I was wondering how it is that they don't actually work. Does the following match your understanding?
- `wait` and `grace-period` are overridden at L235-235 to `false` and `-1` respectively
- `timeout` is always ignored when `wait=false`
- `filenames`, `kustomize` and `recursive` would have to be passed to the Builder created at L425 via `FilenameParam`, but they aren't. So when `DeleteOptions#DeleteResult` runs, the resources iterated over are only the ones given by `Builder#ResourceNames`, i.e. the created objects (good!).
- `force` under the hood actually turns into `grace-period=0`, so our overriding it to `-1` as stated above nullfies this one too. The grace period would show up in the delete request body with -v=8 if this weren't true (it doesn't).

⚠️ I think `cascade` technically does work. If I do `kubectl run -i -t mypod --image=busybox:latest -f base/resources.yaml --rm --pod-running-timeout=1s --cascade=orphan -v=8`, then the object I get back includes `"finalizers":["orphan"]`. That said, this doesn't strike me as useful, since pods don't have dependents by default and this command in any case doesn't create any. But if we do remove it as well, we should use different reasoning. Wdyt?

brianpursley replied:
@KnVerey thanks for tracing through all that.  I should have done that with this PR description!

I agree on all points.  

Also, I'm not sure why I thought cascade was a no-op, but you're right, technically it is used.  I'm not sure what effect it would have in the case of `kubectl run` (there would be no dependent resources?) but the reason for the deprecation is not the same as the other flags.

I'll break it out from the others, with a different comment, and different deprecation message... maybe just `"This flag will be removed in version 1.29"`. 

I'm also not opposed to leaving it, if there is a valid use case. I just can't think of one offhand or how we would explain to people when they should use it.

KnVerey replied:
Yeah I can't really think of one either. Technically you could manually add ownerReferences to objects to cause it to have dependents, but I have no idea why you'd do that for an object created with `run`. Of course, removing the flag doesn't prevent them from doing that, only from changing the deletion policy used with the `rm` option--and `rm` can only be specified when `run` is used interactively (making it extra unlikely this is useful). If we want a reason, maybe "is not relevant for kubectl run"?

KnVerey commented on staging/src/k8s.io/kubectl/pkg/cmd/run/run.go:
```diff
@@ -166,6 +166,23 @@ func NewCmdRun(f cmdutil.Factory, streams genericclioptions.IOStreams) *cobra.Co
+	cmd.Flags().MarkDeprecated("kustomize", "This flag is ignored and will be removed in version 1.29")
+	cmd.Flags().MarkDeprecated("recursive", "This flag is ignored and will be removed in version 1.29")
+	cmd.Flags().MarkDeprecated("timeout", "This flag is ignored and will be removed in version 1.29")
+	cmd.Flags().MarkDeprecated("wait", "This flag is ignored and will be removed in version 1.29")
```
Nit: I built this PR to see how the message actually looks, and apparently it is concatenated to the standard message with a comma:
```
Flag --filename has been deprecated, This flag is ignored and will be removed in version 1.29
Flag --kustomize has been deprecated, This flag is ignored and will be removed in version 1.29
```

Suggestion: `Flag --kustomize has been deprecated, because it is not used by this command. It will be removed in version 1.29`

brianpursley replied:
Yeah, I thought that was a little weird too.  I was following how most of the other deprecation messages were worded, but maybe it is a good time to break from that and try to make it better.  I'll update the messages.

brianpursley replied:
Updated:
```
Flag --filename has been deprecated, because it is not used by this command. It will be removed in version 1.29.
Flag --kustomize has been deprecated, because it is not used by this command. It will be removed in version 1.29.
Flag --cascade has been deprecated, because it is not relevant for this command. It will be removed in version 1.29.
```

Hopefully it will be clear these flags are only going to be removed from THIS command, and not every command.  I think it should be clear enough.

KnVerey replied:
👌

KnVerey reviewed (commented):
/hold 
in case you want to take my nitpicks (up to you)
/lgtm
otherwise

brianpursley committed:
kubectl/pkg/cmd/run: deprecate flags

The --cascade flag, has no practical effect when used, so it is being deprecated.

The following flags, which are unused and ignored by kubectl run,
have been deprecated:

--filename
--force
--grace-period
--kustomize
--recursive
--timeout
--wait

KnVerey commented:
/lgtm 
/hold cancel
